src/compas_ags/ags/graphstatics.py

```diff
@@ -135,18 +133,20 @@ def form_identify_dof(form):
     --------
     >>>
     """
-    vertex_index = form.vertex_index()
-
-    xy = form.vertices_attributes('xy')
-    fixed = [vertex_index[vertex] for vertex in form.fixed()]
-    free = list(set(range(form.number_of_vertices())) - set(fixed))
-    edges = [(vertex_index[u], vertex_index[v]) for u, v in form.edges()]
-    C = connectivity_matrix(edges)
-    E = equilibrium_matrix(C, xy, free)
+    k_i = form.key_index()
+    uv_index = {(u, v): index for index, (u, v) in enumerate(form.edges())}
+    vcount = form.number_of_vertices()
+    fixed = form.leaves()
+    fixed = [k_i[key] for key in fixed]
+    free = list(set(range(vcount)) - set(fixed))
+    ind = [uv_index[key] for key in list(form.edges_where({'is_ind': True}))]
+    edges = [(k_i[u], k_i[v]) for u, v in form.edges()]
+    xy = array(form.vertices_attributes('xy'), dtype=float64).reshape((-1, 2))
+    C = connectivity_matrix(edges, 'array')
+    E = equilibrium_matrix(C, xy, free, 'array')
```

Comment:
the conversion to array is in general not a good idea, but is indeed necessary in this case for the `E` matrix. however, it is therefore handled by the `dof` and `nonpivot` functions. also the conversion of the `C` matrix to appropriate format is handled by the `equilibrium_matrix` function and is now done twice.

this is not important for small matrices, but can be quite costly in more complicated cases...